core/src/main/scala/chisel3/Data.scala

```diff
@@ -435,27 +435,79 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
     _direction = Some(actualDirection)
   }
 
+  private[chisel3] def stringAccessor(chiselType: Option[String], vals: Option[String]): String = {
+    if (this.isLit) {
+      (chiselTypeStrOpt(chiselType),
+        valuesStrOpt(vals)) match {
+        case (Some(a), Some(b)) => s"$a($b)"
+        case (None, None) => ""
+        case (Some(a), _) => s"$a"
+        case (_, Some(b)) => s"$b"
+      }
+    } else {
+      (parentStrOpt,
+        nameStrOpt,
+        subcomponentStrOpt(topBindingOpt.flatMap(_.location)),
+        bindingToStringOpt(),
+        chiselTypeStrOpt(chiselType),
+        valuesStrOpt(vals)) match {
+        //    parent   name     subcomp  binding  type     values
+        case (Some(a), Some(b), Some(c), Some(d), Some(e), Some(f)) => s"$a.$b$c: $d[$e]($f)"
+        case (Some(a), Some(b), Some(c), Some(d), Some(e), None) => s"$a.$b$c: $d[$e]"
+        case (Some(a), None, None, Some(d), Some(e), None) => s"$a: $d[$e]"
+        case (Some(a), None, Some(c), Some(d), Some(e), None) => s"$a.$c: $d[$e]"
+        case (Some(a), Some(b), None, None, Some(e), None) => s"$a.$b: $e"
+        case (_, _, None, None, Some(e), None) => s"$e"
+        case (Some(a), Some(b), _, Some(d), Some(e), _) => s"$a.$b: $d[$e]"
+        case (Some(a), None, None, Some(d), Some(e), Some(f)) => s"$a: $d[$e]($f)"
+        case (None, None, None, None, None, None) => s""
+      }
+    }
+  }
+
```

Comment:
I'd like to see the above logic factored a little bit differently because it really has 2 parts, "name" and "type" that you can construct separately. The current construction creates cases that should be unreachable (eg. when you have a name, you should always have a binding and a parent). There's something to be said about handling all possible cases--if I'm wrong or there's a bug and they do come up, my suggestions will result in an empty String--but I think the more important concern is having the code reflect what can (or is supposed to) actually happen.

I'm thinking something like:
```scala
topBindingOpt match {
  // Just a type, return type
  case None => chiselType  // note this is no longer an `Option` because of changes above
  case Some(topBinding) =>
    val binding: String = _bindingToString(topBinding) // some new method since it's taking just `Binding`
    
    // Note we could replace every `.getOrElse("")` with `.mkString`, it's a stylistic choice
    val name = nameStrOpt.getOrElse("?")
    val subname = subcomponentStrOpt(topBinding.location).getOrElse("")
    val mod = parentStrOpt.map(_ + ".").getOrElse("") // This could even be a `.get`
    s"$mod$name$subname: $binding[$chiselType]"
}
```
I could be wrong, but I think the above will do the same thing while not laying out cases that should be impossible.